spec/grammar.dd

```diff
@@ -680,6 +684,7 @@ $(GNAME ForeachTypeList):
 $(GNAME ForeachType):
     $(GLINK ForeachTypeAttributes)$(OPT) $(GLINK2 declaration, BasicType) $(GLINK2 declaration, Declarator)
     $(GLINK ForeachTypeAttributes)$(OPT) $(I Identifier)
+    $(GLINK ForeachTypeAttributes)$(OPT) $(D alias) $(I Identifier)
```

Comment:
It looks like `static foreach (enum ref alias T; TypeList)` is valid syntax. Is this correct?
Ditto for `foreach (enum ref x; someRange)`.

Reply:
(Should `ForeachTypeAttributes` be allowed before `alias`?)

Reply:
I don't think that separating grammar rules to complicate the parser makes sense. It's good for the parser to be lenient. Semantic analysis should catch it (the resulting error messages are better).

It might however be sensible to not allow `ForeachTypeAttributes` before alias (the reasoning was that global `alias` allows (some) storage classes, but it turns out that local declarations also do not allow storage classes before `alias`), but this pull request documents existing behaviour. You might want to open an enhancement request.

Reply:
I don't think we need to change the compiler if we restructure the grammar so it is a bit harder to produce semantically invalid code.
Yes, you can (and most likely you already do) catch this during semantic analysis, however I feel it would be better for the reader of the spec if the grammar rules match more closely what is semantically valid.
Presently, the reader wouldn't have any clue that `foreach (shared ref enum inout alias x; aggregate)` should be rejected by the compiler - the intended semantics are not specified.